content/blog/r-mpi-part-two.md

```diff
@@ -0,0 +1,298 @@
+
+  send_data <- runif(1 + (end - start))            # My data to send
+  send_ptr <- new_data(send_data)                  # Make C++ data pointer
+
+  recv_data <- if (rank == 0) rep(0, n) else 0     # Only make big array 
+  recv_ptr <- new_data(recv_data)                  # on rank 0
+
+  mpi_time <- -(get_mpi_time())
+  mpi_gather_to_zero(send_ptr, recv_ptr)           # Gather in recv_ptr
+  mpi_time <- mpi_time + get_mpi_time()
+  
+  recv_data <- get_data(recv_ptr)                  # Update R version of recv data
+
+  local_time <- (local_time + get_mpi_time()) - mpi_time
+  
+  if (rank == 0) {
+    message(sprintf("Size %s, mean = %s, local_time = %s, mpi_time = %s, total=%s",
+                    size, mean(recv_data), local_time, mpi_time, local_time + mpi_time))
+  }
+  end_mpi()
+}
+```
+
+We define our `mpi_gather_to_zero` function (which wraps [MPI_Gather](https://www.mpich.org/static/docs/v3.3/www3/MPI_Gather.html)),
+in `src/rmpi.h` :-
+
+```
+void mpi_gather_to_zero(cpp11::external_pointer<std::vector<double>> send,
+                        cpp11::external_pointer<std::vector<double>> recv);
+
+```
+
+and implement it in `src/rmpi.cpp` :-
+```
+[[cpp11::register]]
+void mpi_gather_to_zero(cpp11::external_pointer<std::vector<double>> send,
+                               cpp11::external_pointer<std::vector<double>> recv) {
+
+  std::vector<double> *s = cpp11::as_cpp<cpp11::external_pointer<std::vector<double>>>(send).get();
+  std::vector<double> *r = cpp11::as_cpp<cpp11::external_pointer<std::vector<double>>>(recv).get();
+
+  MPI_Gather(s->data(), s->size(), MPI_DOUBLE,
+             r->data(), s->size(), MPI_DOUBLE, 0, MPI_COMM_WORLD);
+
+}
+```
+
+The `MPI_Gather` call assumes all the nodes will contribute the
+same number of bytes - hence the hack earlier to ensure `n` was exactly
+divisible by `size`. A purer solution would use another function
+[MPI_Gatherv](https://www.mpich.org/static/docs/v3.3/www3/MPI_Gatherv.html),
+that allows processees to send different amounts of data, as long as they all
+agree beforehand how much each process will send.
+
+![Native MPI performance](/img/mpi_2_eg2.png)
+
+And this immediately has the look of something more reasonable. MPI time
+is lower because we're doing less communication, and the scaling is no longer
+messed up by the overheads of getting so much memnory; a larger fraction of
+the run-time is now parallelisable.
+
+We still don't get much efficiency; doubling the processors doesn't half
+the total time - in fact we're nowhere near that. But timing no longer gets
+worse as we add processors! So we just need to be giving the processes more
+work to do.
+
+Lastly, node the run on my desktop is a bit quicker than the cluster job. But
+remember here I'm making poor use of the cluster nodes, only giving them one
+process (one core) each. HPC cluster nodes don't really offer you benefit in
+speed for single core; it's the extra RAM and extra cores that give you the
```

Comment:
Might be nice to mention one or two more realistic examples which would make better use of cluster nodes.

Reply:
I've tried to rewrite this bit...